core/src/apps/wallet/sign_tx/signing.py

```diff
@@ -730,13 +730,18 @@ def output_derive_script(
         elif version == cashaddr.ADDRESS_TYPE_P2SH:
             version = coin.address_type_p2sh
         else:
-            raise ValueError("Unknown cashaddr address type")
+            raise SigningError("Unknown cashaddr address type")
         raw_address = bytes([version]) + data
     else:
-        raw_address = base58.decode_check(o.address, coin.b58_hash)
+        try:
+            raw_address = base58.decode_check(o.address, coin.b58_hash)
+        except ValueError:
+            raise SigningError(FailureType.DataError, "Invalid address")
 
     if address_type.check(coin.address_type, raw_address):
         # p2pkh
+        if not address_type.check(coin.address_type, raw_address):
```

Comment:
this looks like a mistake?
```
if check(address):
   if not check(address):
      ...
```
this and the below checks are covered by the `raise` at the end of the if/elif block?

Reply:
God I am blind. It indeed seems so.

Reply:
that's what reviews are for :)